beep.1
.\" The beep(1) man page
.\"
.\" This man page has been rewritten adhering to the following
.\" documentation: man(7), man-pages(7), tbl(1)
.\"
.TH BEEP 1 "2018-12-29" "Linux" "User commands"
.\"
.\" ====================================================================
.\"
.SH NAME
beep \- beep the pc speaker any number of ways
.\"
.\" ====================================================================
.\"
.SH SYNOPSIS
.B beep
.RB [\| GLOBALS \|]
.RB [\| \-f
.IR FREQ \|]
.RB [\| \-l
.IR LEN \|]
.RB [\| \-r
.IR REPEATS \|]
.RB [\|< \-d|\-D >
.IR DELAY \|]
.RB [\| \-s \|]
.RB [\| \-c \|]
.br
.B beep
.RB [\| GLOBALS \|]
.RI <\| TONE
.IR OPTIONS \|>
.RB [\| \-n|\-\-new \|]
.RI <\| TONE
.IR OPTIONS \|>
.br
.B beep
.RI <\| \-h | \-\-help \|>
.br
.B beep
.RI <\| \-v | \-V | \-\-version \|>
.\"
.\" ====================================================================
.\"
.SH DESCRIPTION
.PP
\fBbeep\fR allows the user to control the PC speaker with precision, allowing different sounds to indicate different events.  While \fBbeep\fR can be run quite happily on the command line, \fBbeep\fR's intended use is within shell/perl scripts, notifying the user when something interesting occurs.  Of course, it has no notion of what \fIIS\fR interesting, but it is good at the notifying part.
.PP
All options have default values, meaning that just typing '\fBbeep\fR' will work.  If an option is specified more than once on the command line, subsequent options override their predecessors.  So '\fBbeep\fR \-f 200 \-f 300' will beep at 300Hz.
.PP
All durations are given in a unit of milliseconds, frequencies as Hertz, and the number of repeats is a dimensionless number.
.\"
.\" ====================================================================
.\"
.SH OPTIONS
.SS "Global options"
.TP
.BI \-e\ DEVICE \fR, \ \fB\-\-device= DEVICE
Explicitly set the device \fBbeep\fR shall use to generate beep tones.  When not given a device explicitly, \fBbeep\fR runs through an internal list of devices until one appears to work.
.TP
.BR \-\-debug ,\  \-\-verbose
Make the \fBbeep\fR program more verbose.
.SS "Tone options"
.TP
.BI \-f\  FREQ
Beep with a tone frequency of \fIFREQ\fR Hz, where 0 < \fIFREQ\fR < 20000.  As a general ballpark, the regular terminal beep is around 750Hz.  For backwards compatibility, you can give \fIFREQ\fR as a floating point number, but \fBbeep\fR will round that to integer values as the kernel APIs expect them.
.TP
.BI \-l\  LEN
Beep for a tone length of \fILEN\fR milliseconds.
.TP
.BI \-r\  REPEATS
Repeat the tone including delays \fIREPEATS\fR times (defaults to 1).
.TP
.BI \-d\  DELAY\fR\ |\  \-D\  DELAY
Specify a delay of \fIDELAY\fR milliseconds between repetitions.  Small letter \fB\-d\fR specifies that this delay should only occur \fIbetween\fR beeps, that is, it should \fInot\fR occur after the last repetition.  Capital letter \fB\-D\fR specifies that the delay should occur after every repetition, \fIincluding\fR the last one.  Normally, \fB\-d\fR is what you want, but if, for example, you are stringing several \fBbeep\fR commands together to play the star wars anthem, you may want control over every delay.
.TP
.BR -n ,\  \-\-new
The \fB\-\-new\fR option allows you to chain multiple beeps together on a single command line.  Each time the \fB\-\-new\FR is used, \fBbeep\fR starts treating all further arguments as though they were for a new beep.  So for example:

    \fBbeep\fR -f 1000 -n -f 2000 -n -f 1500

would produce a sequence of three beeps, the first with a frequency of 1000Hz (and otherwise default values), then a second beep with a frequency of 2000Hz (again, with things like delay and reps being set to their defaults), then a third beep, at 1500Hz.  This is different from specifying a \fB\-r\fR value, since \fB\-r\fR repeats the same beep multiple times, whereas \fB--new\fR allows you to specify different beeps.  After a \fB\-\-new\fR, the new beep is created with all the default values, and any of these can be specified without altering values for preceding (or later) beeps.  See the \fBEXAMPLES\fR section if this managed to confuse you.
.TP
.BR \-s ,\  \-c
Both the \fB\-s\fR and the \fB\-c\fR option put \fBbeep\fR into input processing mode.  \fB\-s\fR tells \fBbeep\fR to read from \fIstdin\fR, and beep after each newline.  \fB\-c\fR tells \fBbeep\fR to beep after every character.  In both cases, the \fBbeep\fR will also echo the input back out to stdout, which makes it easy to slip \fBbeep\fR into a text-processing pipeline.  See the \fBEXAMPLES\fR section.
.SS "Other Actions"
.BR \-h ,\  \-\-help
Display \fBbeep\fR usage info and exit.
.TP
.BR -v ,\  \-V ,\  \-\-version
Display \fBbeep\fR version information and exit.
.\"
.\" ====================================================================
.\"
.SH EXIT STATUS
.PP
When successful,
.B beep
returns with exit code 0.
.PP
Any non-0 exit means
.B beep
has encountered an error and has bailed out.
.\"
.\" ====================================================================
.\"
.SH FILES
The device files
.B beep
uses by default are the following:
.\"
    /dev/input/by-path/platform-pcspkr-event-spkr
    /dev/tty0
    /dev/vc/0
.\"
.\" ====================================================================
.\"
.SH NOTES
.SS "Devices and Permissions"
.PP
When you run \fBbeep\fR without explicitly giving a device to use, \fBbeep\fR tries opening the following devices in sequence until one succeeds:
.IP "\fB/dev/input/by-path/platform-pcspkr-event-spkr\fR" 4
Uses the evdev API, and system administrator can allow access to any set of users and groups using the normal file permissions.
.IP "\fB/dev/tty0\fR" 4
Uses the console API, and the kernel insists you are root or own the TTY.
.IP "\fB/dev/vc/0\fR" 4
Uses the console API, and the kernel insists you are root or own the TTY.
.PP
.B beep
does not support running as setuid root or via
.BR "sudo" "(1)" .
.PP
For non-root uses, the system administrator can set up a
.BR udev "(7)"
rule to allow write access to
.B /dev/input/by-path/platform-pcspkr-event-spkr
for certain users and/or user groups. For details, see the
.B beep
.B INSTALL.md
file.
.\"
.SS "APIs"
.IP "evdev" 4
Uses the input event device driver, which means running
.BR "write" "(2)"
on
.BR /dev/input/by-path/platform-pcspkr-event-spkr .
.IP "console" 4
Uses the ancient console API, which means running the
.B "KIOCSOUND"
.BR "ioctl" "(2)"
on a tty device like
.BR /dev/tty0 or /dev/vc/0.
.\"
.SS "Concurrent Invocations"
Concurrent invocations of
.B beep
are not supported. There is only one tone generator in the PC hardware, and we do not manage access to that shared resource in any way.  So if a first
.B "beep -f 1000 -l 10000"
process starts a long 10 second beep with 1000Hz, and 2 seconds later a second quick
.B "beep -f 2050 -l 100"
with 2050Hz comes in, you will hear that 2050Hz beep for 100ms, and then silence as the quick beep silences the tone generator. Now you will not hear that the first
.B "beep"
process is still running after having waited for the first 2100ms of 10000ms. You will also not that the first
.B "beep"
process will silence the already silent speaker 7900ms later, i.e. 10000ms after the first
.B "beep"
started.
.\"
.SS "Sound Volume"
.PP
The PC speaker hardware has no way to explicitly set the volume of a beep.
.PP
If you have a
.IR "standalone PC" ,
however, chances are you have a piezo beeper connected to the mainboard.  If you prioritize a loud beep over a certain frequency beep, you can choose to beep at a frequency close to the resonance frequency of the piezo beeper.  Typical piezo beepers have a resonance frequency around 2000Hz, so you can test frequencies around that range.  The author's piezo beeper is the loudest at about 2050Hz.
.PP
If you have a
.IR "PC laptop" ,
chances are that your laptop will not have a piezo beeper hardware inside and that it will route the output of the pcspkr circuit to the laptop's internal speakers via the sound card mixer.  In that case, you can and/or must change the mixer level for the beeper using a sound card mixer application.
.\"
.SS "Frequency Table"
.PP
This frequency table might come in hand for translating musical notes to frequencies.  The frequencies are rounded to integer numbers as the kernel driver only works with integers.  The column for
.B "octave 4"
is the octave of middle C.
.\" This table was generated by the gen-freq-table script and
.\" inserted into this man page manually.
.TS
center box;
Cb | Cb s  s  s
Cb | Cb Cb Cb Cb
-  | -  -  -  -
C  | N  N  N  N.
note	octave
name	3	4	5	6	
C	131	262	523	1047	
C#	139	277	554	1109	
D	147	294	587	1175	
D#	156	311	622	1245	
E	165	330	659	1319	
F	175	349	698	1397	
F#	185	370	740	1480	
G	196	392	784	1568	
G#	208	415	831	1661	
A	220	440	880	1760	
A#	233	466	932	1865	
B	247	494	988	1976	
C	262	523	1047	2093	
.TE
.\"
.\" ====================================================================
.\"
.SH BUGS
Report bugs to
.UR https://github.com/ndim/beep/issues
.UE .
.\"
.\" ====================================================================
.\"
.SH EXAMPLES
.PP
At its simplest (yet still effective)

    \fBbeep\fR
.\"
.PP
A more interesting standalone setup

    \fBbeep\fR -f 300.7 -r 2 -d 100 -l 400
.\"
.PP
As part of a log-watching pipeline

    tail -f /var/log/xferlog | grep 'passwd' | \fBbeep\fR -f 1000 -r 5 -s
.\"
.PP
When using -c mode, I recommend using a short -D, and a shorter -l, so that the beeps do not blur together.  Something like this will get you a cheesy 1970's style beep-as-you-type-each-letter effect

    cat file | \fBbeep\fR -c -f 400 -D 50 -l 10
.\"
.PP
A highly contrived example of -n/--new usage

    \fBbeep\fR -f 1000 -r 2 -n -r 5 -l 10 --new

will produce first two 1000Hz beeps, then 5 beeps at the default tone, but only 10ms long each, followed by a third beep using all the default settings (since none are specified).
.\"
.PP See also the \fBFREQUENCY TABLE\fR below.
.\"
.\" ====================================================================
.\"
.SH SEE ALSO
.\" .BR fixme (1),
.\" .BR andme (1),
.BR README.md ,
.BR INSTALL.md ,
.UR "https://github.com/ndim/beep"
.UE


beep.c
/* beep - beep the pc speaker any number of ways
 * Copyright (C) 2000-2010 Johnathan Nightingale
 * Copyright (C) 2010-2013 Gerfried Fuchs
 * Copyright (C) 2013-2018 Hans Ulrich Niedermann
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 */

/*
 * For more documentation on beep, see the beep-usage.txt and
 * beep.1.in files.
 */


#include <errno.h>
#include <fcntl.h>
#include <getopt.h>
#include <limits.h>
#include <math.h>
#include <signal.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/ioctl.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <linux/kd.h>
#include <linux/input.h>

#include "beep-log.h"
#include "beep-usage.h"


/* Use PIT_TICK_RATE value from the kernel. */
#ifndef CLOCK_TICK_RATE
#define CLOCK_TICK_RATE 1193182UL
#endif

const char version_message[] =
    PACKAGE_TARNAME " " PACKAGE_VERSION "\n"
    "Copyright (C) 2002-2016 Johnathan Nightingale\n"
    "Copyright (C) 2013-2018 Hans Ulrich Niedermann\n"
    "Use and Distribution subject to GPL.\n"
    "For information: http://www.gnu.org/copyleft/.\n";

/* Meaningful Defaults */
#define DEFAULT_FREQ       440   /* Middle A */
#define DEFAULT_LENGTH     200   /* milliseconds */
#define DEFAULT_REPS       1
#define DEFAULT_DELAY      100   /* milliseconds */
#define DEFAULT_END_DELAY  NO_END_DELAY
#define DEFAULT_STDIN_BEEP NO_STDIN_BEEP

/* Other Constants */
#define NO_END_DELAY    0
#define YES_END_DELAY   1

#define NO_STDIN_BEEP   0
#define LINE_STDIN_BEEP 1
#define CHAR_STDIN_BEEP 2

typedef struct beep_parms_t {
  unsigned int freq; /* tone frequency (Hz)      */
  unsigned int length;     /* tone length    (ms)      */
  unsigned int reps;       /* # of repetitions         */
  unsigned int delay;      /* delay between reps  (ms) */
  unsigned int end_delay;  /* do we delay after last rep? */
  int stdin_beep; /* are we using stdin triggers?  We have three options:
		     - just beep and terminate (default)
		     - beep after a line of input
		     - beep after a character of input
		     In the latter two cases, pass the text back out again,
		     so that beep can be tucked appropriately into a text-
		     processing pipe.
		  */
  struct beep_parms_t *next;  /* in case -n/--new is used. */
} beep_parms_t;

/* Use an enum and switch statements to have compiler warn us about
 * unhandled cases. */
typedef enum {
	      /* When the beep_type has not been set yet, do nothing */
	      BEEP_TYPE_UNSET   = 0,
	      /* Use the console API */
	      BEEP_TYPE_CONSOLE = 1,
	      /* Use the evdev API */
	      BEEP_TYPE_EVDEV   = 2,
} beep_type_E;

/* Momma taught me never to use globals, but we need something the signal
   handlers can get at.*/
int console_fd = -1;
beep_type_E console_type = BEEP_TYPE_UNSET;
const char *console_device = NULL;


/* We do not know for certain whether perror does strange things with
 * global variables or malloc/free inside its code.
 */
void safe_error_exit(const char *const msg)
{
    const int saved_errno = errno;
    static char strerr_buf[1024];
    const int ret = strerror_r(saved_errno, strerr_buf, sizeof(strerr_buf));
    if (ret != 0) {
        if (write(STDERR_FILENO, "strerror_r error\n",
                  0+0+0+0+strlen("strerror_r error\n"))) {
            /* ignore al write errors */
        }
        _exit(EXIT_FAILURE);
    }
    const size_t msglen = strlen(msg);
    const size_t errlen = strlen(strerr_buf);
    if (write(STDERR_FILENO, msg, msglen)) {
        /* ignore al write errors */
    }
    if (write(STDERR_FILENO, ": ", 2)) {
        /* ignore al write errors */
    }
    if (write(STDERR_FILENO, strerr_buf, errlen)) {
        /* ignore al write errors */
    }
    if (write(STDERR_FILENO, "\n", 1)) {
        /* ignore al write errors */
    }
    _exit(EXIT_FAILURE);
}


void do_beep(unsigned int freq) {
  switch (console_type) {
  case BEEP_TYPE_CONSOLE: if (1) {
      const uintptr_t argp = ((freq != 0) ? (CLOCK_TICK_RATE/freq) : freq) & 0xffff;
      if (-1 == ioctl(console_fd, KIOCSOUND, argp)) {
	/* If we cannot use the sound API, we cannot silence the sound either */
	safe_error_exit("ioctl KIOCSOUND");
      }
    }
    break;
  case BEEP_TYPE_EVDEV: if (1) {
      struct input_event e;

      memset(&e, 0, sizeof(e));
      e.type = EV_SND;
      e.code = SND_TONE;
      e.value = freq;

      if (sizeof(e) != write(console_fd, &e, sizeof(e))) {
	/* If we cannot use the sound API, we cannot silence the sound either */
	safe_error_exit("write EV_SND");
      }
    }
    break;
  case BEEP_TYPE_UNSET:
    /* Do nothing, if this case should ever happen which it should not. */
    break;
  }
}


/* If we get interrupted, it would be nice to not leave the speaker
 * beeping in perpetuity.
 *
 * Everything called from this signal handler must be thread-safe,
 * signal-safe, reentrant including all API functions.  Otherwise, we
 * get another CVE-2018-0492.
 *
 * So we make certain we keep to using the following API calls:
 *
 *   * close(2):      safe
 *   * _exit(2):      safe (which exit(3) is NOT)
 *   * memset(3):     MT-safe
 *   * write(2):      safe
 *   * strerror_r(3): MT-safe
 *   * strlen(3):     MT-safe
 */
void handle_signal(int signum) {
  switch(signum) {
  case SIGINT:
  case SIGTERM:
    if(console_fd >= 0) {
      /* Kill the sound, quit gracefully */
      do_beep(0);
      (void) close(console_fd);
      _exit(signum);
    } else {
      /* Just quit gracefully */
      _exit(signum);
    }
  }
}


/* print usage and leave exit code up to the caller */
void print_usage(void)
{
    fputs(beep_usage, stdout);
}


/* print usage and exit */
void usage_bail(void)
{
    print_usage();
    exit(EXIT_FAILURE);
}


/* whether character is a digit */
int is_digit(char c)
{
  return (('0' <= c) && (c <= '9'));
}


/* whether string consists of at least one digit, and only digits */
int is_number(const char *const str)
{
  if (str[0] == '\0') {
    return 0;
  }
  for (size_t i=0; str[i] != '\0'; ++i) {
    if (!is_digit(str[i])) {
      return 0;
    }
  }
  return 1;
}


/* whether device is on whitelist */
int is_device_whitelisted(const char *const dev)
{
  if (strncmp("/dev/input/", dev, strlen("/dev/input/")) == 0) {
    /* If device name starts with /dev/input/, we can assume evdev
     * input device beeping is wished for and the corresponding
     * device is somewhere in /dev/input/.
     */
    return 1;
  } else if (strcmp("/dev/console", dev) == 0) {
    return 1;
  } else if ((strncmp("/dev/tty", dev, 8) == 0) && is_number(&dev[8])) {
    /* match numbered /dev/tty<N> devices */
    return 1;
  } else if ((strncmp("/dev/vc/", dev, 8) == 0) && is_number(&dev[8])) {
    /* match numbered /dev/vc/<N> devices */
    return 1;
  }
  return 0;
}


/* Parse the command line.  argv should be untampered, as passed to main.
 * Beep parameters returned in result, subsequent parameters in argv will over-
 * ride previous ones.
 * 
 * Currently valid parameters:
 *  "-f <frequency in Hz>"
 *  "-l <tone length in ms>"
 *  "-r <repetitions>"
 *  "-d <delay in ms>"
 *  "-D <delay in ms>" (similar to -d, but delay after last repetition as well)
 *  "-s" (beep after each line of input from stdin, echo line to stdout)
 *  "-c" (beep after each char of input from stdin, echo char to stdout)
 *  "--verbose/--debug"
 *  "-h/--help"
 *  "-v/-V/--version"
 *  "-n/--new"
 *
 * March 29, 2002 - Daniel Eisenbud points out that c should be int, not char,
 * for correctness on platforms with unsigned chars.
 */
void parse_command_line(const int argc, char *const argv[], beep_parms_t *result) {
  int c;

  struct option opt_list[7] = {{"help", 0, NULL, 'h'},
			       {"version", 0, NULL, 'V'},
			       {"new", 0, NULL, 'n'},
			       {"verbose", 0, NULL, 'X'},
			       {"debug", 0, NULL, 'X'},
			       {"device", 1, NULL, 'e'},
			       {0,0,0,0}};
  while((c = getopt_long(argc, argv, "f:l:r:d:D:schvVne:", opt_list, NULL))
	!= EOF) {
    int argval = -1;    /* handle parsed numbers for various arguments */
    float argfreq = -1; 
    switch(c) {      
    case 'f':  /* freq */
      if(!sscanf(optarg, "%f", &argfreq) || (argfreq >= 20000.0f /* ack! */) ||
	 (argfreq <= 0.0f)) {
        usage_bail();
      } else {
	if (result->freq != 0) {
	  log_warning("multiple -f values given, only last one is used.\n");
	}
	result->freq = ((unsigned int)rintf(argfreq));
      }
      break;
    case 'l' : /* length */
        if (!sscanf(optarg, "%d", &argval) || (argval < 0) || (argval > 2100000)) {
            usage_bail();
        } else {
            result->length = argval;
        }
        break;
    case 'r' : /* repetitions */
        if (!sscanf(optarg, "%d", &argval) || (argval < 0) || (argval > 2100000)) {
            usage_bail();
        } else {
            result->reps = argval;
        }
        break;
    case 'd' : /* delay between reps - WITHOUT delay after last beep*/
        if (!sscanf(optarg, "%d", &argval) || (argval < 0) || (argval > 2100000)) {
            usage_bail();
        } else {
            result->delay = argval;
            result->end_delay = NO_END_DELAY;
        }
        break;
    case 'D' : /* delay between reps - WITH delay after last beep */
        if (!sscanf(optarg, "%d", &argval) || (argval < 0) || (argval > 2100000)) {
            usage_bail();
        } else {
            result->delay = argval;
            result->end_delay = YES_END_DELAY;
        }
        break;
    case 's' :
      result->stdin_beep = LINE_STDIN_BEEP;
      break;
    case 'c' :
      result->stdin_beep = CHAR_STDIN_BEEP;
      break;
    case 'v' :
    case 'V' : /* also --version */
        fputs(version_message, stdout);
        exit(EXIT_SUCCESS);
        break;
    case 'n' : /* also --new - create another beep */
      if (result->freq == 0) {
        result->freq = DEFAULT_FREQ;
      }
      result->next = (beep_parms_t *)malloc(sizeof(beep_parms_t));
      result->next->freq       = 0;
      result->next->length     = DEFAULT_LENGTH;
      result->next->reps       = DEFAULT_REPS;
      result->next->delay      = DEFAULT_DELAY;
      result->next->end_delay  = DEFAULT_END_DELAY;
      result->next->stdin_beep = DEFAULT_STDIN_BEEP;
      result->next->next       = NULL;
      result = result->next; /* yes, I meant to do that. */
      break;
    case 'X' : /* --debug / --verbose */
      log_level += 1;
      break;
    case 'e' : /* also --device */
      if (console_device) {
        log_error("You cannot give the --device parameter more than once.\n");
        exit(EXIT_FAILURE);
      }
      if (1) {
	static char realpath_optarg[PATH_MAX+1];
	if (realpath(optarg, realpath_optarg) == NULL) {
	  const int saved_errno = errno;
          log_error("could not run realpath(3) on '%s': %s\n",
                    optarg, strerror(saved_errno));
	  exit(EXIT_FAILURE);
	}
	if (is_device_whitelisted(realpath_optarg)) {
	  console_device = realpath_optarg;
	} else {
          log_error("Not using device '%s'. If you do need this device, please "
                    "report that fact to <https://github.com/spkr-beep/beep/issues>.\n",
                    realpath_optarg);
	  exit(EXIT_FAILURE);
	}
      }
      break;
    case 'h': /* also --help */
        print_usage();
        exit(EXIT_SUCCESS);
        break;
    default:
        usage_bail();
        break;
    }
  }
  if (result->freq == 0) {
      result->freq = DEFAULT_FREQ;
  }
}


void play_beep(beep_parms_t parms) {
  unsigned int i; /* loop counter */

  log_verbose("%d times %d ms beeps (%d delay between, "
              "%d delay after) @ %d Hz\n",
              parms.reps, parms.length, parms.delay, parms.end_delay, parms.freq);

  /* Beep */
  for (i = 0; i < parms.reps; i++) {                    /* start beep */
    do_beep(parms.freq);
    /* Look ma, I'm not ansi C compatible! */
    usleep(1000U*parms.length);                          /* wait...    */
    do_beep(0);                                         /* stop beep  */
    if (parms.end_delay || ((i+1) < parms.reps)) {
       usleep(1000U*parms.delay);                        /* wait...    */
    }
  }                                                     /* repeat.    */
}


/* Open only character device special file (with race condition).
 *
 * We check whether this is a character device special file before
 * opening as for some devices, opening has an effect and we can avoid
 * this effect for those devices here.
 *
 * We still need to make sure that the file we have actually opened
 * actually is a character device special file after we have actually
 * opened it.
 */
int open_chr(const char *filename, int flags)
{
  struct stat sb;
  if (-1 == stat(filename, &sb)) {
    return -1;
  }
  if (S_ISCHR(sb.st_mode)) {
    return open(filename, flags);
  } else {
    log_error("console file '%s' is not a character device special file\n",
              filename);
    exit(EXIT_FAILURE);
  }
}


/* If stdout is a TTY, print a bell character to stdout as a fallback. */
void fallback_beep(void)
{
  /* Printing '\a' can only beep if we print it to a tty */
  if (isatty(STDOUT_FILENO)) {
    putc('\a', stdout);
  }
}


int main(const int argc, char *const argv[]) {
  log_init(argc, argv);

  char sin[4096], *ptr;

  /* Bail out if running setuid or setgid.
   *
   * It is near impossible to make beep setuid-safe:
   *
   *   * We open files for writing, and may even write to them.
   *
   *   * Checking the device file with realpath leaks information.
   *
   * So we refuse running setuid or setgid.
   */
  if ((getuid() != geteuid()) || (getgid() != getegid())) {
    log_error("Running setuid or setgid, "
              "which is not supported for security reasons.\n");
    log_error("Set up permissions for the pcspkr evdev device file instead.\n");
    exit(EXIT_FAILURE);
  }

  /* Bail out if running under sudo.
   *
   * For the reasoning, see the setuid comment above.
   */
  if (getenv("SUDO_COMMAND") || getenv("SUDO_USER") || getenv("SUDO_UID") || getenv("SUDO_GID")) {
    log_error("Running under sudo, "
          "which is not supported for security reasons.\n");
    log_error("Set up permissions for the pcspkr evdev device file instead.\n");
    exit(EXIT_FAILURE);
  }

  /* Parse command line */
  beep_parms_t *parms = (beep_parms_t *)malloc(sizeof(beep_parms_t));
  if (NULL == parms) {
    perror("malloc");
    exit(EXIT_FAILURE);
  }
  parms->freq       = 0;
  parms->length     = DEFAULT_LENGTH;
  parms->reps       = DEFAULT_REPS;
  parms->delay      = DEFAULT_DELAY;
  parms->end_delay  = DEFAULT_END_DELAY;
  parms->stdin_beep = DEFAULT_STDIN_BEEP;
  parms->next       = NULL;

  parse_command_line(argc, argv, parms);

  /* Try opening a console device */
  if (console_device) {
    console_fd = open_chr(console_device, O_WRONLY);
  } else {
    static char *console_device_list[] =
      { "/dev/input/by-path/platform-pcspkr-event-spkr",
	"/dev/tty0",
	"/dev/vc/0",
      };
    for (size_t i=0; i<(sizeof(console_device_list)/sizeof(console_device_list[0])); ++i) {
      if ((console_fd = open_chr(console_device_list[i], O_WRONLY)) != -1) {
	console_device = console_device_list[i];
	break;
      }
    }
  }

  if (console_fd == -1) {
    const int saved_errno = errno;
    log_error("Could not open %s for writing: %s\n",
              ((console_device != NULL) ? console_device : "console device"),
              strerror(saved_errno));
    /* Output the only beep we can, in an effort to fall back on usefulness */
    fallback_beep();
    exit(EXIT_FAILURE);
  }

  /* Verify that console_fd is actually a character device special file */
  if (1) {
    struct stat sb;
    if (-1 == fstat(console_fd, &sb)) {
      perror("fstat");
      exit(EXIT_FAILURE);
    }
    if (S_ISCHR(sb.st_mode)) {
      /* GOOD: console_fd is a character device special file. Use it. */
    } else {
      /* BAD: console_fd is not a character device special file. Do
       * not use it any further, and certainly DO NOT WRITE to it.
       */
      log_error("opened console '%s' is not a character special device\n",
                console_device);
      exit(EXIT_FAILURE);
    }
  }

  /* Determine the API supported by the opened console device */
  if (ioctl(console_fd, EVIOCGSND(0)) != -1) {
    log_verbose("Using BEEP_TYPE_EVDEV on '%s'\n", console_device);
    console_type = BEEP_TYPE_EVDEV;
  } else if (ioctl(console_fd, KIOCSOUND, 0) >= 0) {
    /* turning off the beeps should be a safe way to check for API support */
    log_verbose("Using BEEP_TYPE_CONSOLE on '%s'\n", console_device);
    console_type = BEEP_TYPE_CONSOLE;
  } else {
    log_error("console device '%s' does not support either API\n",
              console_device);
    /* Output the only beep we can, in an effort to fall back on usefulness */
    fallback_beep();
    exit(EXIT_FAILURE);
  }

  /* At this time, we know what API to use on which device, and we do
   * not have to fall back onto printing '\a' any more.
   */

  /* Memory barrier. All globals have been set up, so we make sure the
   * values are actually written to memory.  Only then do we install
   * the signal handlers.
   *
   * TBD: Use C11 atomic_signal_fence() instead of "__asm__ volatile"?
   */
  __asm__ volatile ("" : : : "memory");

  /* After all the initialization has happened and the global
   * variables used to communicate with the signal handlers have
   * actually been set up properly, we can finally install the signal
   * handlers. As we do not start making any noises, there is no need
   * to install the signal handlers any earlier.
   */
  signal(SIGINT,  handle_signal);
  signal(SIGTERM, handle_signal);

  /* this outermost while loop handles the possibility that -n/--new has been
     used, i.e. that we have multiple beeps specified. Each iteration will
     play, then free() one parms instance. */
  while(parms) {
    beep_parms_t *next = parms->next;

    if(parms->stdin_beep) {
      /* in this case, beep is probably part of a pipe, in which case POSIX 
	 says stdin and out should be fuly buffered.  This however means very 
	 laggy performance with beep just twiddling it's thumbs until a buffer
	 fills. Thus, kill the buffering.  In some situations, this too won't 
	 be enough, namely if we're in the middle of a long pipe, and the 
	 processes feeding us stdin are buffered, we'll have to wait for them,
	 not much to  be done about that. */
      setvbuf(stdin, NULL, _IONBF, 0);
      setvbuf(stdout, NULL, _IONBF, 0);
      while(fgets(sin, 4096, stdin)) {
	if(parms->stdin_beep==CHAR_STDIN_BEEP) {
	  for(ptr=sin;*ptr;ptr++) {
	    putchar(*ptr);
	    fflush(stdout);
	    play_beep(*parms);
	  }
	} else {
	  fputs(sin, stdout);
	  play_beep(*parms);
	}
      }
    } else {
      play_beep(*parms);
    }

    /* Junk each parms struct after playing it */
    free(parms);
    parms = next;
  }

  (void) close(console_fd);
  console_fd = -1;

  return EXIT_SUCCESS;
}

/*
 * Local Variables:
 * c-basic-offset: 4
 * indent-tabs-mode: nil
 * End:
 */
